File: raspberryPi/product/serial_debug_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
シリアル通信デバッグテスト（PySerial + termios 版）
16バイト固定受信（VMIN=16, VTIME=5 相当）
"""

import sys
import time
import signal
import termios
import serial

running = True

def handle_sigint(signum, frame):
    """Ctrl+C でループを抜ける"""
    global running
    running = False

def test_serial_ports():
    """利用可能なシリアルポートをテスト"""
    ports_to_test = [
        "/dev/ttyUSB0",
        "/dev/ttyUSB1",
        "/dev/ttyAMA0",
        "/dev/serial0",
        "/dev/ttyS0"
    ]
    print("🔍 利用可能なシリアルポートを検索中…")
    for port in ports_to_test:
        try:
            ser = serial.Serial(
                port=port,
                baudrate=9600,
                bytesize=serial.EIGHTBITS,
                parity=serial.PARITY_EVEN,
                stopbits=serial.STOPBITS_ONE,
                timeout=1
            )
            print(f"✅ {port}: 接続成功")
            ser.close()
        except Exception as e:
            print(f"❌ {port}: {e}")

def monitor_serial(port="/dev/ttyUSB0"):
    """16バイト固定受信でシリアルをモニタリング"""
    global running
    try:
        # ブロッキング読み込みにするため timeout=None
        ser = serial.Serial(
            port=port,
            baudrate=9600,
            bytesize=serial.EIGHTBITS,
            parity=serial.PARITY_EVEN,
            stopbits=serial.STOPBITS_ONE,
            timeout=None
        )

        # termios で VMIN=16, VTIME=5 を設定
        fd = ser.fileno()
        attrs = termios.tcgetattr(fd)
        # attrs[6] は c_cc 配列
        attrs[6][termios.VMIN]  = 16   # 最低受信バイト数
        attrs[6][termios.VTIME] = 5    # 0.5秒（単位はデシ秒）
        termios.tcsetattr(fd, termios.TCSANOW, attrs)

        print(f"📡 シリアルモニタリング開始: {port}")
        print("    設定: 9600bps, 8bit, Even parity, 1 stop bit")
        print("    VMIN=16, VTIME=5 (0.5秒)")
        print("    Ctrl+C で終了\n")

        signal.signal(signal.SIGINT, handle_sigint)
        signal.signal(signal.SIGTERM, handle_sigint)

        last_activity = time.time()

        while running:
            # 16バイト読んで返ってくる（VMIN/VTIME に従う）
            data = ser.read(16)
            if data:
                ts = time.strftime("%H:%M:%S")
                hexstr = data.hex().upper()
                ascstr = ''.join(
                    chr(b) if 32 <= b <= 126 else '.' for b in data
                )
                print(f"[{ts}] 受信 ({len(data)} バイト)")
                print(f"  HEX  : {hexstr}")
                print(f"  ASCII: {ascstr}\n")
                last_activity = time.time()
            else:
                # タイムアウト（VTIME）で n == 0
                if time.time() - last_activity > 10:
                    print(f"[{time.strftime('%H:%M:%S')}] 待機中… (データなし)\n")
                    last_activity = time.time()

        print("\n🛑 モニタリング終了")

    except Exception as e:
        print(f"❌ エラー: {e}")
    finally:
        if 'ser' in locals() and ser.is_open:
            ser.close()

def main():
    if len(sys.argv) > 1:
        if sys.argv[1] == "test":
            test_serial_ports()
        else:
            monitor_serial(sys.argv[1])
    else:
        print("使用方法:")
        print("  python3 serial_debug_test.py test          # ポート検索")
        print("  python3 serial_debug_test.py /dev/ttyUSB0  # モニタリング\n")
        test_serial_ports()
        print()
        monitor_serial()

if __name__ == "__main__":
    main()

File: raspberryPi/product/serial_debug_test.c
// serial_debug_test.c
// シリアル通信デバッグテスト（POSIX termios版）
// ビルド例: gcc serial_debug_test.c -o serial_debug_test

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <termios.h>
#include <time.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>

static volatile int running = 1;

// シグナルハンドラー (Ctrl+C でループを抜ける)
void handle_sigint(int sig) {
    (void)sig;
    running = 0;
}

// シリアルポートを開いて termios 設定を行う
//   portname: "/dev/ttyUSB0" など
//   nonblock: テストモードなら 1（ノンブロッキング）、モニタリングなら 0
//   戻り値: ファイルディスクリプタ (>=0) or エラーで -1
int open_serial(const char *portname, int nonblock) {
    int flags = O_RDWR | O_NOCTTY | (nonblock ? O_NONBLOCK : 0);
    int fd = open(portname, flags);
    if (fd < 0) return -1;

    // モニタリング時は VTIME/VMIN でタイムアウト制御するため
    if (!nonblock) {
        // ブロッキングモードに戻す
        fcntl(fd, F_SETFL, 0);

        struct termios tio;
        if (tcgetattr(fd, &tio) < 0) {
            close(fd);
            return -1;
        }
        // ボーレート設定
        cfsetispeed(&tio, B9600);
        cfsetospeed(&tio, B9600);
        // 8bit, Even parity, 1 stop bit
        tio.c_cflag &= ~CSIZE;
        tio.c_cflag |= CS8;
        tio.c_cflag |= PARENB;   // parity enable
        tio.c_cflag &= ~PARODD;  // even
        tio.c_cflag &= ~CSTOPB;  // 1 stop bit
        // raw モード
        tio.c_lflag = 0;
        tio.c_iflag = 0;
        tio.c_oflag = 0;
        // フレーム長(16)を指定。最大0.5秒待機。
        tio.c_cc[VMIN]  = 16;
        tio.c_cc[VTIME] = 5;
        if (tcsetattr(fd, TCSANOW, &tio) < 0) {
            close(fd);
            return -1;
        }
    }

    return fd;
}

// モニタリング処理
void monitor_serial(const char *port) {
    int fd = open_serial(port, 0);
    if (fd < 0) {
        fprintf(stderr, "❌ %s を開けません: %s\n", port, strerror(errno));
        return;
    }

    printf("📡 シリアルモニタリング開始: %s\n", port);
    printf("    設定: 9600bps, 8bit, Even parity, 1 stop bit\n");
    printf("    Ctrl+C で終了\n\n");

    signal(SIGINT, handle_sigint);
    signal(SIGTERM, handle_sigint);

    time_t last_activity = time(NULL);

    unsigned char buf[256];
    while (running) {
        int n = read(fd, buf, sizeof(buf));
        if (n < 0) {
            // 読み取りエラーなら少し待って再試行
            usleep(100000);
            continue;
        }
        if (n > 0) {
            // 受信あり
            char ts[16];
            time_t now = time(NULL);
            struct tm lt;
            localtime_r(&now, &lt);
            strftime(ts, sizeof(ts), "%H:%M:%S", &lt);

            // HEX 表示
            char hexstr[256*2+1] = {0};
            for (int i = 0; i < n; i++) {
                sprintf(hexstr + i*2, "%02X", buf[i]);
            }
            // ASCII 表示
            char ascstr[256+1] = {0};
            for (int i = 0; i < n; i++) {
                unsigned char b = buf[i];
                ascstr[i] = (b >= 32 && b <= 126) ? b : '.';
            }

            printf("[%s] 受信 (%dバイト)\n", ts, n);
            printf("  HEX  : %s\n", hexstr);
            printf("  ASCII: %s\n\n", ascstr);

            last_activity = now;
        } else {
            // タイムアウト (VTIME) で n == 0
            time_t now = time(NULL);
            if (now - last_activity > 10) {
                char ts[16];
                struct tm lt;
                localtime_r(&now, &lt);
                strftime(ts, sizeof(ts), "%H:%M:%S", &lt);
                printf("[%s] 待機中... (データなし)\n", ts);
                last_activity = now;
            }
        }
        // read 自体が最大 0.1s ブロックするので追加 Sleep は不要
    }

    printf("\n🛑 モニタリング終了\n");
    close(fd);
}

// ポート検索テスト
void test_serial_ports(void) {
    const char* ports[] = {
        "/dev/ttyUSB0",
        "/dev/ttyUSB1",
        "/dev/ttyAMA0",
        "/dev/serial0",
        "/dev/ttyS0",
    };
    size_t cnt = sizeof(ports)/sizeof(ports[0]);

    printf("🔍 利用可能なシリアルポートを検索中...\n");
    for (size_t i = 0; i < cnt; i++) {
        int fd = open_serial(ports[i], 1);
        if (fd >= 0) {
            printf("✅ %s: 接続成功\n", ports[i]);
            close(fd);
        } else {
            printf("❌ %s: %s\n", ports[i], strerror(errno));
        }
    }
}

// エントリポイント
int main(int argc, char *argv[]) {
    if (argc > 1) {
        if (strcmp(argv[1], "test") == 0) {
            test_serial_ports();
            return 0;
        } else {
            monitor_serial(argv[1]);
            return 0;
        }
    }

    // 引数なし時
    printf("使用方法:\n");
    printf("  %s test          # ポート検索\n", argv[0]);
    printf("  %s /dev/ttyUSB0  # モニタリング\n\n", argv[0]);
    test_serial_ports();
    printf("\n");
    monitor_serial("/dev/ttyUSB0");
    return 0;
}
